File: estateman-frontend/src/services/realtors.ts
import { api } from './api';

export interface Realtor {
  id: number;
  realtor_id: string;
  user_id: number;
  level: 'junior' | 'senior' | 'team_lead';
  status: 'active' | 'inactive' | 'suspended';
  total_clients: number;
  active_deals: number;
  total_commissions: number;
  monthly_target: number;
  monthly_earned: number;
  rating: number;
  specialties: string[];
  location?: string;
  license_number?: string;
  license_expiry?: string;
  commission_rate: number;
  split_percentage: number;
  join_date: string;
  created_at: string;
  updated_at?: string;
  // Team management fields
  team_id?: number;
  manager_id?: number;
  // Performance tracking fields
  ytd_commissions?: number;
  avg_deal_size?: number;
  conversion_rate?: number;
  client_satisfaction?: number;
  response_time_hours?: number;
}

export interface Commission {
  id: number;
  realtor_id: number;
  sale_price: number;
  commission_rate: number;
  gross_commission: number;
  split_percentage: number;
  net_commission: number;
  status: string;
  paid_date?: string;
  payment_method?: string;
  created_at: string;
}

export interface RealtorAnalytics {
  total_realtors: number;
  active_realtors: number;
  total_commissions: number;
  average_rating: number;
  top_performers: Array<{
    id: number;
    name: string;
    realtor_id: string;
    total_commissions: number;
    level: string;
  }>;
}

export interface CommissionAnalytics {
  total_commission: number;
  pending_payouts: number;
  this_month: number;
  commission_rate: number;
  recent_commissions: Array<{
    id: number;
    realtor_name: string;
    sale_price: number;
    net_commission: number;
    status: string;
    created_at: string;
  }>;
}

export interface RealtorCreateData {
  name: string;
  email: string;
  phone?: string;
  level?: Realtor['level'];
  specialties?: string[];
  location?: string;
  license_number?: string;
  license_expiry?: string;
  monthly_target?: number;
  commission_rate?: number;
  split_percentage?: number;
}

export interface RealtorUpdateData {
  level?: Realtor['level'];
  status?: Realtor['status'];
  specialties?: string[];
  location?: string;
  license_number?: string;
  license_expiry?: string;
  monthly_target?: number;
  commission_rate?: number;
  split_percentage?: number;
  team_id?: number;
  manager_id?: number;
  ytd_commissions?: number;
  avg_deal_size?: number;
  conversion_rate?: number;
  client_satisfaction?: number;
  response_time_hours?: number;
}

export interface CommissionCreateData {
  realtor_id: number;
  sale_price: number;
  commission_rate: number;
  split_percentage: number;
  transaction_id?: number;
  property_id?: number;
  client_id?: number;
}

class RealtorsService {
  // Realtor methods
  async getRealtors(params?: {
    skip?: number;
    limit?: number;
    level?: Realtor['level'];
    status?: Realtor['status'];
    search?: string;
  }): Promise<Realtor[]> {
    const response = await api.get('/realtors/', { params });
    return response.data;
  }

  async getRealtor(realtorId: number): Promise<Realtor> {
    const response = await api.get(`/realtors/${realtorId}`);
    return response.data;
  }

  async createRealtor(data: RealtorCreateData): Promise<Realtor> {
    const response = await api.post('/realtors/', data);
    return response.data;
  }

  async updateRealtor(realtorId: number, data: RealtorUpdateData): Promise<Realtor> {
    const response = await api.put(`/realtors/${realtorId}`, data);
    return response.data;
  }

  async deleteRealtor(realtorId: number): Promise<void> {
    await api.delete(`/realtors/${realtorId}`);
  }

  async getRealtorAnalytics(): Promise<RealtorAnalytics> {
    const response = await api.get('/realtors/analytics/overview');
    return response.data;
  }

  async getRealtorPerformance(realtorId: number): Promise<any> {
    const response = await api.get(`/realtors/performance/${realtorId}`);
    return response.data;
  }

  // Commission methods
  async getCommissions(params?: {
    skip?: number;
    limit?: number;
    realtor_id?: number;
    status?: string;
  }): Promise<Commission[]> {
    const response = await api.get('/realtors/commissions/', { params });
    return response.data;
  }

  async createCommission(data: CommissionCreateData): Promise<Commission> {
    const response = await api.post('/realtors/commissions/', data);
    return response.data;
  }

  async getCommissionAnalytics(): Promise<CommissionAnalytics> {
    const response = await api.get('/realtors/commissions/analytics/overview');
    return response.data;
  }

  async calculateCommission(salePrice: number, rate: number, split: number): Promise<any> {
    const response = await api.post('/realtors/commissions/calculate', null, {
      params: { sale_price: salePrice, rate, split }
    });
    return response.data;
  }

  async getRealtorsDropdown(search?: string): Promise<Array<{id: number, name: string, realtor_id: string}>> {
    try {
      console.log('Fetching realtors dropdown with search:', search);
      const response = await api.get('/realtors/dropdown', { params: search ? { search } : {} });
      console.log('Realtors dropdown response:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Realtors dropdown error:', {
        message: error?.message,
        status: error?.response?.status,
        statusText: error?.response?.statusText,
        data: error?.response?.data,
        url: error?.config?.url,
        method: error?.config?.method,
        params: error?.config?.params
      });
      throw error;
    }
  }

  // Enhanced realtor profile methods
  async getRealtorProfile(realtorId: number): Promise<any> {
    const response = await api.get(`/realtors/${realtorId}/profile`);
    return response.data;
  }

  async updateRealtorProfile(realtorId: number, data: any): Promise<Realtor> {
    const response = await api.put(`/realtors/${realtorId}/profile`, data);
    return response.data;
  }

  async getRealtorTeamInfo(realtorId: number): Promise<any> {
    const response = await api.get(`/realtors/${realtorId}/team`);
    return response.data;
  }

  async getRealtorPerformanceMetrics(realtorId: number): Promise<any> {
    const response = await api.get(`/realtors/${realtorId}/performance-metrics`);
    return response.data;
  }

  // Events methods
  async getEvents(): Promise<any[]> {
    const response = await api.get('/realtors/events');
    return response.data;
  }

  async registerForEvent(eventId: number): Promise<any> {
    const response = await api.post(`/realtors/events/${eventId}/register`);
    return response.data;
  }

  // Marketing methods
  async getMarketingMaterials(): Promise<any[]> {
    const response = await api.get('/realtors/marketing/materials');
    return response.data;
  }

  async getRealtorCampaigns(realtorId: number): Promise<any[]> {
    const response = await api.get(`/realtors/${realtorId}/marketing/campaigns`);
    return response.data;
  }

  // Leaderboard methods
  async getLeaderboard(limit: number = 10): Promise<any[]> {
    const response = await api.get('/realtors/leaderboard', { params: { limit } });
    return response.data;
  }

  async getRealtorRanking(realtorId: number): Promise<any> {
    const response = await api.get(`/realtors/${realtorId}/ranking`);
    return response.data;
  }

  // Additional methods for portal data
  async getAnnouncements(): Promise<any[]> {
    const response = await api.get('/announcements');
    return response.data;
  }

  async getRealtorLeads(realtorId: number): Promise<any[]> {
    const response = await api.get(`/realtors/${realtorId}/leads`);
    return response.data;
  }

  async getRealtorSalesData(realtorId: number): Promise<any> {
    const response = await api.get(`/realtors/${realtorId}/sales-data`);
    return response.data;
  }

  async getCommissionHistory(realtorId: number): Promise<any[]> {
    const response = await api.get(`/realtors/${realtorId}/commission-history`);
    return response.data;
  }

  async getNetworkMembers(realtorId: number): Promise<any[]> {
    const response = await api.get(`/realtors/${realtorId}/network-members`);
    return response.data;
  }

  async getNotifications(realtorId: number): Promise<any[]> {
    const response = await api.get(`/realtors/${realtorId}/notifications`);
    return response.data;
  }

  async invitePartner(realtorId: number, inviteData: {
    name: string;
    email: string;
    phone?: string;
    message?: string;
  }): Promise<any> {
    const response = await api.post(`/realtors/${realtorId}/invite-partner`, inviteData);
    return response.data;
  }

  async createCampaign(realtorId: number, campaignData: {
    name: string;
    type: string;
    target_audience: string;
    message: string;
    schedule_date?: string;
  }): Promise<any> {
    const response = await api.post(`/realtors/${realtorId}/marketing/campaigns`, campaignData);
    return response.data;
  }
}

export const realtorsService = new RealtorsService();

File: estateman-frontend/estateman-backend/app/api/v1/endpoints/realtors.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from app.api.deps import get_db, get_current_user
from app.models.user import User
from app.models.realtor import RealtorLevel, RealtorStatus
from app.schemas.realtor import (
    RealtorCreate, RealtorUpdate, RealtorResponse,
    CommissionCreate, CommissionResponse, CommissionUpdate,
    TransactionCreate, TransactionResponse,
    RealtorAnalytics, CommissionAnalytics,
    LeadCreate, LeadResponse, NetworkStats, NetworkMember,
    MarketingMaterial, CampaignCreate, CampaignResponse,
    EventResponse, LeaderboardEntry, RealtorProfile,
    ClientPortfolio, ActivityLog
)
from app.services.realtor import RealtorService, CommissionService

router = APIRouter()

# Realtor endpoints
@router.post("/", response_model=RealtorResponse)
def create_realtor(
    realtor_data: RealtorCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.create_realtor(realtor_data)

@router.get("/", response_model=List[RealtorResponse])
def get_realtors(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    level: Optional[RealtorLevel] = None,
    status: Optional[RealtorStatus] = None,
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    realtors = service.get_realtors(skip, limit, level, status, search)
    return realtors or []

@router.get("/{realtor_id}", response_model=RealtorResponse)
def get_realtor(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    realtor = service.get_realtor(realtor_id)
    if not realtor:
        return {"message": "The realtor you are trying to access does not exist or has been removed"}
    return realtor

@router.put("/{realtor_id}", response_model=RealtorResponse)
def update_realtor(
    realtor_id: int,
    realtor_data: RealtorUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    realtor = service.update_realtor(realtor_id, realtor_data)
    if not realtor:
        return {"message": "The realtor you are trying to update does not exist or has been removed"}
    return realtor

@router.delete("/{realtor_id}")
def delete_realtor(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    if not service.delete_realtor(realtor_id):
        return {"message": "The realtor you are trying to delete does not exist or has already been removed"}
    return {"message": "Realtor deleted successfully"}

@router.get("/analytics/overview", response_model=RealtorAnalytics)
def get_realtor_analytics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_analytics()

@router.get("/dropdown")
def get_realtors_dropdown(
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    realtors = service.get_realtors(0, 100, None, None, search)
    if not realtors:
        return []
    
    result = []
    for r in realtors:
        if r.user_id and r.realtor_id:
            name = f"{r.user.first_name} {r.user.last_name}" if r.user else f"Realtor {r.realtor_id}"
            result.append({
                "id": r.user_id, 
                "name": name, 
                "realtor_id": r.realtor_id
            })
    return result

@router.get("/performance/{realtor_id}")
def get_realtor_performance(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    performance = service.get_realtor_performance(realtor_id)
    if not performance:
        return {"message": "No performance data available for this realtor - they may be new or inactive"}
    return performance

# Commission endpoints
@router.post("/commissions/", response_model=CommissionResponse)
def create_commission(
    commission_data: CommissionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.create_commission(commission_data)

@router.get("/commissions/", response_model=List[CommissionResponse])
def get_commissions(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    realtor_id: Optional[int] = None,
    status: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    commissions = service.get_commissions(skip, limit, realtor_id, status)
    return commissions or []

@router.get("/commissions/analytics/overview", response_model=CommissionAnalytics)
def get_commission_analytics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.get_commission_analytics()

@router.post("/commissions/calculate")
def calculate_commission(
    sale_price: float,
    rate: float,
    split: float,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.calculate_commission(sale_price, rate, split)

# Team Management endpoints
@router.get("/team/{team_lead_id}")
def get_team_members(
    team_lead_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.get_team_members(team_lead_id)

@router.get("/performance/detailed/{realtor_id}")
def get_detailed_performance(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.get_detailed_performance(realtor_id)

@router.put("/performance/update/{realtor_id}")
def update_performance_metrics(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    realtor = service.update_performance_metrics(realtor_id)
    if not realtor:
        raise HTTPException(status_code=404, detail="Realtor not found")
    return {"message": "Performance metrics updated successfully", "realtor_id": realtor.id}

# Enhanced Profile endpoints
@router.get("/{realtor_id}/profile")
def get_realtor_profile(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_profile(realtor_id)

@router.get("/{realtor_id}/clients")
def get_realtor_clients(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_clients(realtor_id)

@router.get("/{realtor_id}/activities")
def get_realtor_activities(
    realtor_id: int,
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_activities(realtor_id, limit)

# Network/MLM endpoints
@router.get("/{realtor_id}/network/stats")
def get_network_stats(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_network_stats(realtor_id)

@router.get("/{realtor_id}/network/downline")
def get_network_downline(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_network_downline(realtor_id)

@router.get("/{realtor_id}/network/tree")
def get_network_tree(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_network_tree(realtor_id)

# Lead Management endpoints
@router.get("/{realtor_id}/leads")
def get_realtor_leads(
    realtor_id: int,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_leads(realtor_id, skip, limit)

@router.get("/{realtor_id}/leads/stats")
def get_leads_stats(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_leads_stats(realtor_id)

# Marketing endpoints
@router.get("/marketing/materials")
def get_marketing_materials(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_marketing_materials()

@router.get("/{realtor_id}/marketing/campaigns")
def get_realtor_campaigns(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_campaigns(realtor_id)

# Events endpoints
@router.get("/events")
def get_events(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_events()

@router.post("/events/{event_id}/register")
def register_for_event(
    event_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.register_for_event(current_user.id, event_id)

# Leaderboard endpoints
@router.get("/leaderboard")
def get_leaderboard(
    limit: int = Query(10, ge=1, le=50),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_leaderboard(limit)

@router.get("/{realtor_id}/ranking")
def get_realtor_ranking(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = RealtorService(db)
    return service.get_realtor_ranking(realtor_id)

# Enhanced Commission endpoints
@router.get("/{realtor_id}/commissions/history")
def get_commission_history(
    realtor_id: int,
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.get_commission_history(realtor_id, skip, limit)

@router.get("/{realtor_id}/commissions/breakdown")
def get_commission_breakdown(
    realtor_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.get_commission_breakdown(realtor_id)

@router.get("/{realtor_id}/commissions/statement")
def generate_commission_statement(
    realtor_id: int,
    start_date: str,
    end_date: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    from datetime import datetime
    service = CommissionService(db)
    start = datetime.fromisoformat(start_date)
    end = datetime.fromisoformat(end_date)
    return service.generate_commission_statement(realtor_id, start, end)

@router.get("/{realtor_id}/commissions/forecast")
def forecast_commission(
    realtor_id: int,
    months_ahead: int = Query(3, ge=1, le=12),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.forecast_commission(realtor_id, months_ahead)

@router.put("/commissions/{commission_id}/approve")
def approve_commission(
    commission_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    commission = service.approve_commission_payout(commission_id, current_user.id)
    if not commission:
        raise HTTPException(status_code=404, detail="Commission not found")
    return {"message": "Commission approved successfully", "commission_id": commission_id}

@router.get("/{realtor_id}/tax-forms/{tax_year}")
def generate_tax_form(
    realtor_id: int,
    tax_year: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    service = CommissionService(db)
    return service.generate_tax_form_data(realtor_id, tax_year)